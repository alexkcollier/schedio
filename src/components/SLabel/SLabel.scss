.label {
  @include font-size(5);
  display: block;
  font-weight: $font-weight-bold;
  text-transform: uppercase;

  &--inline {
    align-items: center;
    display: inline-flex;
    flex-direction: row;
    justify-content: flex-start;

    .input,
    .textarea,
    .select {
      margin-left: spacing();
    }

    &.label--reversed {
      flex-direction: row-reverse;
    }
  }

  &--fullwidth {
    width: 100%;
  }

  &:not(:last-child) {
    margin-right: spacing();
  }

  &:not(:first-child) {
    margin: spacing('loose') 0;
  }

  &__optional {
    @include font-size(6);
    color: color('night', 'lightest');
    font-weight: $font-weight-light;
    text-transform: none;
  }
}
