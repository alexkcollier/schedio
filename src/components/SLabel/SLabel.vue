<template>
  <label :class="classList" class="label">
    <!-- Wrapped to keep optional on the right when reversed -->
    <span>
      {{ label }}
      <span v-if="isOptional" class="label__optional">&mdash;&nbsp;Optional</span>
    </span>
    <slot/>
  </label>
</template>

<script>
export default {
  name: 'SLabel',

  props: {
    label: {
      type: String,
      required: true
    },

    isOptional: {
      type: Boolean,
      default: true
    },

    isInline: {
      type: Boolean,
      default: false
    },

    isFullwidth: {
      type: Boolean,
      default: false
    },

    isReversed: {
      type: Boolean,
      default: false
    }
  },

  computed: {
    classList() {
      return {
        'label--inline': this.isInline,
        'label--fullwidth': this.isFullwidth,
        'label--reversed': this.isReversed
      }
    }
  },

  mounted() {
    if ((this.isFullwidth || this.isReversed) && !this.isInline) {
      console.error('`isFullwidth` and `isReversed` require `isInline`')
    }
  }
}
</script>
