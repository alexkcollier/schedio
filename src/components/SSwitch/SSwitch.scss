.switch {
  @include switch;

  &--base ~ &__check {
    @include switch-size('base');
  }

  &--medium ~ &__check {
    @include switch-size('medium');
  }

  &--large ~ &__check {
    @include switch-size('large');
  }
}
