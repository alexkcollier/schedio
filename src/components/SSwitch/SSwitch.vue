<template>
  <SLabel
    :for="id"
    :label="label"
    :is-reversed="isReversed"
    :is-required="required"
    :hide-optional="hideOptional"
    is-inline
  >
    <input
      :id="id"
      v-bind="$attrs"
      :name="name"
      :required="required"
      :class="{ [`switch--${size}`]: size }"
      class="switch"
      type="checkbox"
      v-on="listeners"
    >
    <span
      :class="{ 'switch__check--invalid': isInvalid }"
      class="switch__check"
    />
  </SLabel>
</template>

<script>
import InputControl from '@/mixins/InputControl.mixin'

const sizes = ['small', 'medium', 'large']

export default {
  name: 'SSwitch',

  mixins: [InputControl],

  props: {
    size: {
      type: String,
      default: '',
      validator: s => !s || (sizes.includes(s) || console.error(`\`size\` must be one of ${sizes}`))
    }
  }
}
</script>
