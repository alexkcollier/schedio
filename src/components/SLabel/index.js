import SLabel from './SLabel.vue'

export { SLabel }

export default { install: Vue => Vue.component(SLabel.name, SLabel) }
