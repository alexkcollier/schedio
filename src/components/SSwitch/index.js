import SSwitch from './SSwitch.vue'

export { SSwitch }

export default {
  install: Vue => Vue.component(SSwitch.name, SSwitch),
}
