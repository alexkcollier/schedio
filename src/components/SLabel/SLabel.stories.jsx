import { SHeading } from '@/components/SHeading'
import { SLabel } from '@/components/SLabel'
import PropList from '@@/docs/components/PropList'
import { boolean, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'

storiesOf('Components/Inputs.SLabel', module)
  .addParameters({ jest: 'SLabel' })
  .add('Label', () => {
    return {
      props: {
        props: {
          default: () => ({
            label: text('Label text', 'Label text', 'Required Props'),
            isFullwidth: boolean('isFullwidth', false, 'Optional Props'),
            isInline: boolean('isInline', false, 'Optional Props'),
            isReversed: boolean('isReversed', false, 'Optional Props'),
            isRequired: boolean('iRrequired', false, 'Optional Props'),
            hideOptional: boolean('hide-optional', false, 'Optional Props')
          })
        },

        defaultSlot: {
          default: text('Default', 'SomeInputComponent', 'Slots')
        }
      },
      render(h) {
        const { props, defaultSlot } = this.$props
        return (
          <div>
            <SHeading level="1">Label</SHeading>
            <p>
              <code>SLabel</code> provides labels for all form inputs. It is included automatically
              for all input components in this library, and is designed to wrap components.
            </p>
            <p>
              By default, if a label is for an optional field, it will append '- Optional' to the
              label. This text can be hidden with the <code>hide-optional</code> prop.
            </p>

            <SHeading level="2">Example</SHeading>
            <SLabel {...{ props }}>
              <div style="text-transform: none; font-weight: 300;">
                <code>{'<' + defaultSlot + ' />'}</code>
              </div>
            </SLabel>

            <PropList component={SLabel} />
          </div>
        )
      }
    }
  })
