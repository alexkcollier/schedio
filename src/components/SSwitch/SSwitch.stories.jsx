import { SHeading } from '@/components/SHeading'
import { SSwitch } from '@/components/SSwitch'
import { withAttrsAsProps, withUnboundAttrs } from '@/mixins/stories/form-fields'
import PropList from '@@/docs/components/PropList'
import { boolean, select } from '@storybook/addon-knobs'

export default {
  title: 'Components/Inputs.SSwitch',

  parameters: {
    jest: 'SSwitch',
  },
}

export const switchStory = () => ({
  props: {
    props: {
      default: () => ({
        ...withAttrsAsProps({
          id: 'toggle-switch',
          name: 'toggle-switch',
          label: 'Toggle Switch',
        }),
        isReversed: boolean('is-reversed', false, 'Optional Props'),
        size: select('size', ['', 'base', 'medium', 'large'], '', 'Optional Props'),
      }),
    },
    attrs: {
      default: withUnboundAttrs({ value: 'toggle-switch' }),
    },
  },
  render (h) {
    const { props, attrs } = this.$props
    return (
      <div>
        <SHeading level="1">Switch</SHeading>
        <p>Switches toggle a single setting off or on.</p>
        <SHeading level="2">Example</SHeading>
        <SSwitch {...{ props }} {...{ attrs }} />

        {SSwitch.props && <PropList component={SSwitch} />}
      </div>
    )
  },
})

switchStory.story = {
  name: 'Switch',
}
