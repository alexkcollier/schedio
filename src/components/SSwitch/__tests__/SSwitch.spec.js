import { mount } from '@vue/test-utils'
import SSwitch from '@/components/SSwitch/SSwitch.vue'
import SLabel from '@/components/SLabel/SLabel.vue'

describe('SSwitch.vue', () => {
  const sizes = ['small', 'medium', 'large']
  const inputEvent = jest.fn()
  const spy = jest.spyOn(global.console, 'error').mockImplementation(() => {})
  const defaultProps = {
    id: 'toggleSwitch',
    name: 'toggleSwitch',
    label: 'Checkbox'
  }
  let wrapper
  let toggleSwitch

  beforeEach(() => {
    wrapper = mount(SSwitch, {
      propsData: { ...defaultProps },
      listeners: {
        input: inputEvent
      }
    })

    toggleSwitch = wrapper.find('#toggleSwitch')
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('requires required props', () => {
    // spies on `console.error` without calling `console.error`
    // mounting component without required props
    mount(SSwitch)
    expect(spy).toBeCalled()
    expect(spy.mock.calls[0][0]).toContain('[Vue warn]: Missing required prop')
  })

  it('can be checked and unchecked', () => {
    // check
    wrapper.trigger('click')
    expect(toggleSwitch.element.checked).toBe(true)

    // uncheck
    wrapper.trigger('click')
    expect(toggleSwitch.element.checked).toBe(false)

    // check that events were emitted correctly
    expect(wrapper.emitted('input')).toEqual([[true], [false]])
  })

  it('can be a required input', () => {
    wrapper.setProps({ required: true })
    expect(toggleSwitch.element.required).toBe(true)
  })

  it('can be invalid', () => {
    wrapper.setProps({ isInvalid: true })
    expect(toggleSwitch.element.validity.valid).toBe(false)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can be disabled', () => {
    const disabledWrapper = mount(SSwitch, {
      propsData: {
        id: 'toggleSwitch',
        name: 'toggleSwitch',
        label: 'Checkbox'
      },
      attrs: {
        disabled: true
      }
    })

    // verifies disabled attr is passed through to toggleSwitch input element
    expect(disabledWrapper.attributes('disabled')).toBeUndefined()
    expect(disabledWrapper.find('#toggleSwitch').element.disabled).toBeTruthy()
  })

  it('can be reversed', () => {
    wrapper.setProps({ isReversed: true })
    expect(wrapper.find(SLabel).props('isReversed')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  sizes.forEach(size => {
    it(`can be size ${size}`, () => {
      const sizeWrapper = mount(SSwitch, {
        propsData: {
          ...defaultProps,
          size
        }
      })
      expect(spy).not.toBeCalled()
      expect(sizeWrapper.html()).toMatchSnapshot()
    })
  })

  it('validates size', () => {
    mount(SSwitch, {
      propsData: {
        ...defaultProps,
        size: 'this is not a size'
      }
    })
    expect(spy).toBeCalledTimes(2)
  })

  it('can hide the optional tag', () => {
    wrapper.setProps({ showOptional: false, required: true })
    expect(wrapper.text()).not.toContain('Optional')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
