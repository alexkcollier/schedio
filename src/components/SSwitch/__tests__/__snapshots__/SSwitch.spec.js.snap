// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSwitch.vue can be invalid 1`] = `
<label class="label label--inline" for="toggleSwitch"><span>
    Checkbox
    <!----></span> <input id="toggleSwitch" name="toggleSwitch" required="required" type="checkbox" class="switch"> <span class="switch__check switch__check--invalid"></span></label>
`;

exports[`SSwitch.vue can be reversed 1`] = `
<label class="label label--inline label--reversed" for="toggleSwitch"><span>
    Checkbox
    <!----></span> <input id="toggleSwitch" name="toggleSwitch" required="required" type="checkbox" class="switch"> <span class="switch__check"></span></label>
`;

exports[`SSwitch.vue can be size large 1`] = `
<label class="label label--inline" for="toggleSwitch"><span>
    Checkbox
    <!----></span> <input id="toggleSwitch" name="toggleSwitch" required="required" type="checkbox" class="switch switch--large"> <span class="switch__check"></span></label>
`;

exports[`SSwitch.vue can be size medium 1`] = `
<label class="label label--inline" for="toggleSwitch"><span>
    Checkbox
    <!----></span> <input id="toggleSwitch" name="toggleSwitch" required="required" type="checkbox" class="switch switch--medium"> <span class="switch__check"></span></label>
`;

exports[`SSwitch.vue can be size small 1`] = `
<label class="label label--inline" for="toggleSwitch"><span>
    Checkbox
    <!----></span> <input id="toggleSwitch" name="toggleSwitch" required="required" type="checkbox" class="switch switch--small"> <span class="switch__check"></span></label>
`;

exports[`SSwitch.vue renders correctly 1`] = `
<label class="label label--inline" for="toggleSwitch"><span>
    Checkbox
    <!----></span> <input id="toggleSwitch" name="toggleSwitch" required="required" type="checkbox" class="switch"> <span class="switch__check"></span></label>
`;
